stdlib/public/runtime/StaticBinaryELF.cpp

```diff
@@ -0,0 +1,354 @@
+//===--- StaticBinaryELF.cpp ----------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// Parse a static ELF binary to implement lookupSymbol() address lookup.
+//
+//===----------------------------------------------------------------------===//
+
+#if defined(__ELF__) && defined(__linux__)
+
+#include "ImageInspection.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/Optional.h"
+#include "llvm/ADT/StringRef.h"
+#include "swift/Basic/Lazy.h"
+#include <cassert>
+#include <string>
+#include <vector>
+#include <elf.h>
+#include <fcntl.h>
+#include <pthread.h>
+#include <unistd.h>
+#include <linux/limits.h>
+#include <sys/auxv.h>
+#include <sys/mman.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+using namespace std;
+using namespace llvm;
+
+#if __POINTER_WIDTH__ == 64
+#define ELFCLASS ELFCLASS64
+typedef Elf64_Ehdr Elf_Ehdr;
+typedef Elf64_Shdr Elf_Shdr;
+typedef Elf64_Phdr Elf_Phdr;
+typedef Elf64_Addr Elf_Addr;
+typedef Elf64_Word Elf_Word;
+typedef Elf64_Sym Elf_Sym;
+typedef Elf64_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF64_ST_TYPE(x)
+#else
+#define ELFCLASS ELFCLASS32
+typedef Elf32_Ehdr Elf_Ehdr;
+typedef Elf32_Shdr Elf_Shdr;
+typedef Elf32_Phdr Elf_Phdr;
+typedef Elf32_Addr Elf_Addr;
+typedef Elf32_Word Elf_Word;
+typedef Elf32_Sym Elf_Sym;
+typedef Elf32_Section Elf_Section;
+#define ELF_ST_TYPE(x) ELF32_ST_TYPE(x)
+#endif
```

Comment:
I think that using a templated trait type might be better.

```
template <typename Bits_>
struct ELFTraits;

template <>
struct ELFTraits<4> {
  using Header = Elf32_Ehdr;
  using SectionHeader = Elf32_Shdr;
  ...
  static inline constexpr uint32_t symbol_type(value) {
    return ELF32_ST_TYPE(value);
  }
};

template <>
struct ELFTraits<8> {
  ...
};

using ELF = ELFTraits<__SIZEOF_POINTER__>;
```

Reply:
Maybe we could reuse LLVM's ELFObjectFile headers here?

Reply:
Thats a good idea.

Reply:
I used the `ELFType` in the end. I had a look at `ELFObjectFile` but it wanted to read in the whole file rather than just the sections that were needed.